applications/FluidDynamicsApplication/custom_elements/d_vms_dem_coupled.cpp

```diff
@@ -313,24 +443,43 @@ template< class TElementData >
 void DVMSDEMCoupled<TElementData>::CalculateStabilizationParameters(
     const TElementData& rData,
     const array_1d<double,3> &Velocity,
-    double &TauOne,
-    double &TauTwo) const
+    Matrix &TauOne,
+    double &TauTwo)
 {
     const double h = rData.ElementSize;
     const double density = this->GetAtCoordinate(rData.Density,rData.N);
     const double viscosity = this->GetAtCoordinate(rData.EffectiveViscosity,rData.N);
     constexpr double mTauC1 = DVMS<TElementData>::mTauC1;
     constexpr double mTauC2 = DVMS<TElementData>::mTauC2;
+    Matrix permeability = this->GetAtCoordinate(rData.Permeability, rData.N);
+    Matrix inv_permeability = ZeroMatrix(Dim, Dim);
+    Matrix non_diag_tau_one = ZeroMatrix(Dim, Dim);
+    Matrix inv_tau = ZeroMatrix(Dim, Dim);
+    Matrix I = IdentityMatrix(Dim, Dim);
+    Matrix eigen_values_matrix, eigen_vectors_matrix;
+    Matrix inv_eigen_matrix = ZeroMatrix(Dim, Dim);
+
+    double det_permeability = MathUtils<double>::Det(permeability);
+    MathUtils<double>::InvertMatrix(permeability, inv_permeability, det_permeability, -1.0);
 
     double velocity_norm = Velocity[0]*Velocity[0];
     for (unsigned int d = 1; d < Dim; d++)
         velocity_norm += Velocity[d]*Velocity[d];
     velocity_norm = std::sqrt(velocity_norm);
 
-    double inv_tau = mTauC1 * viscosity / (h*h) + density * ( 1.0/rData.DeltaTime + mTauC2 * velocity_norm / h );
-    TauOne = 1.0/inv_tau;
-    TauTwo = viscosity + density * mTauC2 * velocity_norm * h / mTauC1;
+    inv_tau = (mTauC1 * viscosity / (h*h) + density * ( 1.0/rData.DeltaTime + mTauC2 * velocity_norm / h )) * I + viscosity * inv_permeability;
 
+    double det_inv_tau = MathUtils<double>::Det(inv_tau);
+    MathUtils<double>::InvertMatrix(inv_tau, non_diag_tau_one, det_inv_tau, -1.0);
+
+    MathUtils<double>::GaussSeidelEigenSystem<Matrix, Matrix>(non_diag_tau_one, eigen_vectors_matrix, eigen_values_matrix, 1.0e-16, 20);
```

Comment:
I think that setting the machine precision as tolerance is strictly enforcing you to always do 20 iterations. Also note that the `GaussSeidelEigenSystem` returns a `bool` indicating the convergence. This allows issuing a warning in case it's `false`.